src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -211,46 +230,74 @@ private void Stop()
                 IPEndPoint remoteEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.RemoteAddress);
+                string targetHost = string.Empty;   // compat with SslStream
+                if (connectionInfo.ServerNameLength > 0 && connectionInfo.ServerName != IntPtr.Zero)
+                {
+                    // TBD We should figure out what to do with international names.
+                    targetHost = Marshal.PtrToStringAnsi(connectionInfo.ServerName, connectionInfo.ServerNameLength);
+                }
 
-                connectionHandle = new SafeMsQuicConnectionHandle(evt.Data.NewConnection.Connection);
+                SafeMsQuicConfigurationHandle? connectionConfiguration = state.ConnectionConfiguration;
 
-                uint status = MsQuicApi.Api.ConnectionSetConfigurationDelegate(connectionHandle, state.ConnectionConfiguration);
-                QuicExceptionHelpers.ThrowIfFailed(status, "ConnectionSetConfiguration failed.");
+                if (connectionConfiguration == null)
+                {
+                    Debug.Assert(state.AuthenticationOptions.ServerCertificateSelectionCallback != null);
+                    try
+                    {
+                        // ServerCertificateSelectionCallback is synchronous. We will call it as needed when building configuration
+                        connectionConfiguration = SafeMsQuicConfigurationHandle.Create(state.ConnectionOptions, state.AuthenticationOptions, targetHost);
+                    }
+                    catch (Exception ex)
+                    {
+                        if (NetEventSource.Log.IsEnabled())
+                        {
+                            NetEventSource.Error(state, $"[Listener#{state.GetHashCode()}] Exception occurred during creating configuration in connection callback: {ex}");
+                        }
+                    }
+
+                    if (connectionConfiguration == null)
+                    {
+                        // We don't have safe handle yet so MsQuic will cleanup new connection.
+                        return MsQuicStatusCodes.InternalError;
+                    }
+                }
 
-                var msQuicConnection = new MsQuicConnection(localEndPoint, remoteEndPoint, connectionHandle, state.RemoteCertificateRequired, state.RevocationMode, state.RemoteCertificateValidationCallback);
-                msQuicConnection.SetNegotiatedAlpn(connectionInfo.NegotiatedAlpn, connectionInfo.NegotiatedAlpnLength);
+                connectionHandle = new SafeMsQuicConnectionHandle(evt.Data.NewConnection.Connection);
 
-                if (!state.AcceptConnectionQueue.Writer.TryWrite(msQuicConnection))
+                uint status = MsQuicApi.Api.ConnectionSetConfigurationDelegate(connectionHandle, connectionConfiguration);
+                if (MsQuicStatusHelper.SuccessfulStatusCode(status))
```

Comment:
Why we don't `QuicExceptionHelpers.ThrowIfFailed` here as we do everywhere else?
This would be the only place in the whole codebase where don't do that, so why? What's different here?

Reply:
now, this is long story. The main reason is because this is executed from native callback on msquic thread. 
However, I made update with something I wanted for long time - validation callback failures will be now propagated to ConnectAsync() on client and then visible to caller. (this is ongoing pain on SslStream)
(more work to probably come)